### PR TITLE
Fix nested relations

### PR DESCRIPTION
**Pull request by pxpm**
Please use #4116 to test this. That should be merged here for this to properly work. 

It's separated so we can discuss separate issues in separate PR.

WHAT DONT WORK 4 SURE:
- nested repeatables (so all relations that could use repeatable don't work, they work as selects/direct fields).

### CODE REVIEW

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 1]
```diff
@@ -107,21 +107,19 @@ private function createRelationsForItem($item, $formattedRelations)
             $relationType = $relationDetails['relation_type'];
 
             switch ($relationType) {
-                case 'BelongsTo':
```
Comment (pxpm) [inline src/app/Library/CrudPanel/Traits/Create.php:110]: Don needed anymore, BelongsTo relations are always saved along with entry.

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 2]
```diff
@@ -107,21 +107,19 @@ private function createRelationsForItem($item, $formattedRelations)
...
-                    if ($modelInstance != null) {
-                        $relation->associate($modelInstance)->save();
-                    } else {
-                        $relation->dissociate()->save();
-                    }
-                    break;
                 case 'HasOne':
                 case 'MorphOne':
-                        $modelInstance = $this->createUpdateOrDeleteOneToOneRelation($relation, $relationMethod, $relationDetails);
+                        if (is_null($relationDetails['values'])) {
```
Comment (pxpm) [inline src/app/Library/CrudPanel/Traits/Create.php:112]: Basically moved the function over here, but removed some things that covered cases that we don't have anymore.

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 3]
```diff
@@ -134,20 +132,21 @@ private function createRelationsForItem($item, $formattedRelations)
...
+                            if (isset($value[$relationMethod])) {
+                                $relationValues[$value[$relationMethod]] = Arr::except($value, $relationMethod);
+                            }
                         }
                     }
 
                     // if there is no relation data, and the values array is single dimensional we have
                     // an array of keys with no aditional pivot data. sync those.
-                    if (empty($relationValues)) {
+                    if (empty($relationValues) && ! is_multidimensional_array($values)) {
```
Comment (pxpm) [inline src/app/Library/CrudPanel/Traits/Create.php:149]: We had this before, adding back as it is needed.

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 4]
```diff
@@ -225,29 +173,41 @@ private function createUpdateOrDeleteOneToOneRelation($relation, $relationMethod
...
-            $removed_entries = $model_instance->where($relation_foreign_key, $item->{$relation_local_key});
+            $removed_entries = $modelInstance->where($relationForeignKey, $item->{$relationLocalKey});
 
-            return $this->handleManyRelationItemRemoval($model_instance, $removed_entries, $relationDetails, $relation_foreign_key);
+            return $this->handleManyRelationItemRemoval($modelInstance, $removed_entries, $relationDetails, $relationForeignKey);
         }
-        // we add the new values into the relation
-        $model_instance->whereIn($model_instance->getKeyName(), $relation_values)
-            ->update([$relation_foreign_key => $item->{$relation_local_key}]);
+        // we add the new values into the relation, if it is HasMany we only update the foreign_key,
```
Comment (pxpm) [inline src/app/Library/CrudPanel/Traits/Create.php:187]: This segment adds support for MorphMany selectable. I was only accounting for HasMany.

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 5]
```diff
@@ -338,33 +298,125 @@ private function getRelationDetailsFromInput($input)
...
+                        $relationDetails = $this->geFieldDetailsForRelationSaving($subfield, $input, $relationDetails, $field);
+                    }
+                }
+            }
+        }
+
+        return $relationDetails;
+    }
+
+    private function geFieldDetailsForRelationSaving($field, $input, $relationDetails, $parent = false)
```
Comment (pxpm) [inline src/app/Library/CrudPanel/Traits/Create.php:318]: This is the HEAVY change in this PR. 

We will discuss this better tomorrow, if you have other stuff to look at, please ignore code review in this function as I am going to refactor it in the morning.

File: src/app/Library/CrudPanel/Traits/Update.php [hunk 6]
```diff
@@ -113,10 +113,6 @@ private function getModelAttributeValueFromRelationship($model, $field)
                 if (! isset($field['subfields'])) {
                     return $related_model->{$relation_method};
                 }
-                // we want to exclude the "self pivot field" since we already have it.
```
Comment (pxpm) [inline src/app/Library/CrudPanel/Traits/Update.php:116]: We added `->getAttributes()` for the model in a previous PR, so we don't rely anymore on the pivot fields to get the values from database, we just get the pivot values.

File: src/resources/views/crud/fields/relationship/select.blade.php [hunk 7]
```diff
@@ -51,6 +51,10 @@
 
     $current_value = !is_array($current_value) ? $current_value->toArray() : $current_value;
 
+    if($field['name'] === 'universes' && $field['pivot'] === false) {
```
Comment (pxpm) [inline src/resources/views/crud/fields/relationship/select.blade.php:54]: I need to remove this!

File: src/resources/views/crud/fields/relationship/select.blade.php [hunk 8]
```diff
@@ -51,6 +51,10 @@
...
+        //dd($current_value, $field);
```
Comment (pxpm) [inline src/resources/views/crud/fields/relationship/select.blade.php:55]: and remove this too!

File: src/resources/views/crud/fields/repeatable.blade.php [hunk 9]
```diff
@@ -1,7 +1,9 @@
 {{-- REPEATABLE FIELD TYPE --}}
 
 @php
-  $field['value'] = old_empty_or_null($field['name'], []) ??  $field['value'] ?? $field['default'] ?? [];
+  $fieldNameForRepeatable = $field['real_name'] ?? $field['name'];
```
Comment (pxpm) [inline src/resources/views/crud/fields/repeatable.blade.php:4]: small baby step to support repeatable over repeatable. 

Atleast they don't break the page anymore but still don't properly work:

image

File: src/resources/views/crud/fields/repeatable.blade.php [hunk 10]
```diff
@@ -250,7 +252,7 @@ function setupRepeatableDeleteButtonEvent(repeatable_group) {
             delete_button.click(function(){
 
                 let $repeatableElement = $(this).closest('.repeatable-element');
-                let container = $('[data-repeatable-holder='+$($repeatableElement).attr('data-repeatable-identifier')+']')
+                let container = $('[data-repeatable-holder="'+$($repeatableElement).attr('data-repeatable-identifier')+'"]')
```
Comment (tabacitu) [inline src/resources/views/crud/fields/repeatable.blade.php:255]: All these tiny changes for quotes and stuff? The renaming of the variables from snake_case to `camelCase`? **Next time, please do them as a separate PR or a separate commit.** I understand if doing a PR is too time-consuming, but a separate commit? Come on, that's easy 😀  

Let me explain WHY that is useful when we're working in teams, on big PRs. And why that's the way I want us to work together going forward.

If these small changes were a separate commit, I could have [cherry-picked](https://mattstauffer.com/blog/how-to-merge-only-specific-commits-from-a-pull-request/) these small changes, merge them into 4.2. They're inconsequential, I could have just gotten them out of the way, and then we would be left with the "heavy stuff" in this PR, to work on it and comment the hell out of it for days. But no, we have both small (that would get merged without even testing) and big stuff (that will definitely need a lot of back-and-forth, refactoring, etc), all in one big PR. No bueno 😀  In those cases, what you want to do is separate them... but we can't do that either, because look at the git history:

Screenshot 2022-01-25 at 11 37 56

What could I cherry-pick from these? 😅

So let's make a few new rules, to force us to be better collaborators:
- if it's a _tiny change_ (the reviewer will merge without testing), create a _separate commit_ for it;
- if it's a _small change_ (that the reviewer can test and merge separately), create a _separate PR_ for it;
- if it's a _medium change_ ... I don't know, we'll see;
- if it's a _big change_ (will definitely be some back-and-forth on it), then break it down into tiny/small/medium changes; 
    - create an issue for the big change, breaking it down into separate smaller problems that need to be tackled;
    - if there is any "low hanging fruit" that will be inconsequential and can be submitted as small PRs, do that first;
    - decide what problem should be tackled first (usually the one the others depend on);
    - create a PR for that first problem; submit the PR and wait for review; 
    - after review or green light, create a PR for the second one;
    - so on and so forth;

What's an easy way to do that? **Let's write good commit messages**. From now on:
- "refactor" is not acceptable as a commit message;
- "wip" is not acceptable as a commit message;

Every commit has to explain itself. To make it clear WHAT that commit does. Even better, **let's try never to use "_and_" in commit messages**. That will further force us to make small commits, that can easily be cherry-picked. Let's start paying attention to our git history, I think it will greatly improve our collaboration.

Another rule: **If it takes more than one day to do, it’s not a task. It’s a collection of tasks.** Treat it as such. Break it down into smaller pieces. Create an issue for it, then treat the first problem alone. Get feedback on it, then progress. See “big change” above. Then you'll never be forced to write a commit message like "wip", because tasks are manageable. Personally, I do more than that - I consider a task something that I can do in less than 3h. Something like that. If it'll take more than 3h to do... it's definitely a collection of tasks.

Sorry for bothering you with this now @pxpm , but this year I've made it a mission to improve a few things in how we collaborate. These small things will add up to a HUGE difference, I promise. And you will _love_ the results! We'll talk more in a meeting. But I'm excited about this too - I think it'll be super-good - for you, me and Backpack!

Until then, rest assured that if I took the time to write this out (especially today), it's worth taking seriously, and doing that in the future. Feel free to argue with me if you don't agree with any of the rules. But I do want to create a series of rules we stick by, from now on. Even create a document for them. Even post them on twitter or smth.

Comment (tabacitu) [reply src/resources/views/crud/fields/repeatable.blade.php:255]: For now, nothing to do about the above, other than read and "commit" (😅 ) to doing it. It's already too late for this PR.

File: src/resources/views/crud/fields/repeatable.blade.php (hunk 9 above)
Comment (tabacitu) [reply src/resources/views/crud/fields/repeatable.blade.php:4]: Ugh... 



I know we've talked about doing something like this. But we've talked about doing it for ALL fields, when we refactor them to objects. Having separate `real_name` and `name` for some fields, and not others... this will become a mess in no time.

Why is this needed more exactly?

Comment (tabacitu) [reply src/resources/views/crud/fields/repeatable.blade.php:4]: Also, if we do do this (which is a big if).

Why do we need a separate variable for it? Why not do:
```suggestion
  $field['name'] = $field['real_name'] ?? $field['name'];
```

File: src/resources/views/crud/fields/repeatable.blade.php [hunk 11]
```diff
@@ -1,7 +1,9 @@
 {{-- REPEATABLE FIELD TYPE --}}
 
 @php
-  $field['value'] = old_empty_or_null($field['name'], []) ??  $field['value'] ?? $field['default'] ?? [];
+  $fieldNameForRepeatable = $field['real_name'] ?? $field['name'];
+  $field['value'] = old_empty_or_null($fieldNameForRepeatable, []) ??  $field['value'] ?? $field['default'] ?? [];
```
Comment (tabacitu) [inline src/resources/views/crud/fields/repeatable.blade.php:5]: If we did the above, this would still be:
```suggestion
  $field['value'] = old_empty_or_null($field['name'], []) ??  $field['value'] ?? $field['default'] ?? [];
```

But of course, it might have other implications, if `$field['name']` is used below for different purposes (and I assume it is).

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 5 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:318]: Ok I'll wait. Just from seeing the HUGE id/else statement I can tell that the interior of this should probably become 2 separate functions or something.

**Review by tabacitu — CHANGES_REQUESTED** [review]
I'm super-happy to hear it's all done in this one Pedro! 🎉🎉🎉

But as it is right now... this PR cannot be merged. It does too much, in too many places. I'll wait for you to refactor the bit you said you want to refactor... then we should find a way to break this down into smaller chunks that CAN be merged.

I love the fact that you added comments here in the PR, where the code itself isn't self-explanatory, they definitely help. But we're still a long way from having this in a point where I say "_ok, I understand what this does and how, I'm certain this won't have any negative implications in other parts_".

If you can make PRs out of the small stuff here (changing single-quotes to double-quotes, changing variable names for snake_case to camel_case), let do that first, merge those to 4.2. At least this PR will become smaller. I think that's the only place we can start right now.

**Comment by tabacitu** [comment]
You know what we can do? After your refactor, let's have a meeting and re-do this PR, following the rules I've outlined above. I think it'll be a good way to test the rules, practice the rules, and see the difference they would make.

Let's do a pair-programming session, where we redo this PR in a way I would be happy to merge it.

File: src/resources/views/crud/fields/inc/repeatable_row.blade.php [hunk 12]
```diff
@@ -17,31 +17,26 @@
...
+                        // since repeatable changes names both in JS and PHP we need to make sure we have a name
+                        // that corresponds to the full relation that we can later convert into brakets to use in repeatable
+                        // and also use it as key in repeatable if defined (repeatable over repeatable)
+                        $subfield['real_name'] = $field['name'].'.'.$subfield['name'];
+                        $subfield['name'] = $field['name'].'['.$repeatable_row_key.']['.$subfield['name'].']';
+                        
+                    }else{
+                        $subfield['value'] = \Arr::get($row, $subfield['name']);
+
+                        $subfield['name'] = $field['name'].'['.$repeatable_row_key.']['.Str::replace('.', '][', $subfield['name']).']';
```
Comment (tabacitu) [inline src/resources/views/crud/fields/inc/repeatable_row.blade.php:39]: Is this change needed now, in this PR? Or is it related to https://github.com/Laravel-Backpack/CRUD/pull/4116/files ? Or was this a previous bug, that we didn't know about?

**Comment by tabacitu** [comment]
I believe https://github.com/Laravel-Backpack/CRUD/pull/4127 is a complete replacement for this. So I've slapped the "WON'T" badge on this. Am I correct @pxpm ? If so, let's please close this PR and delete the branch. We can always undelete it later.
